### PR TITLE
SPT-4113: Support for App and Record Revisions, along with history fields

### PR DESCRIPTION
**Pull request by jondavidford**
Expands upon Michael Butler's previous PR here: https://github.com/swimlane/swimlane-python/pull/56

I initially just started by adding tests for the history field, but realized that the right way to implement this was to add adapters and APIResource objects for app and record revisions. This way it is easier to implement resource caching and testing is simpler and more robust. Additionally, we get the ability to fetch revisions directly from App and Record objects using the following syntax:

`app_object.revisions.get(3) # gets revision number 3 for this app` (the only method to get app revisions is individually like this)

`record_object.revisions.get() # gets all revisions for this record` (the only method to get record revisions is all at once like this)

Currently, resource caching is only implemented for AppRevisions. The cache decorator methods are set up to cache individual objects, so caching an entire response of RecordRevisions (the only way record revisions are retrieved currently) didn't make much sense given the current architecture. The way AppRevisions are cached is somewhat hacky, since we don't retrieve them by the same unique identifier we use to cache them. But I think it is robust enough for now.

### CODE REVIEW

File: swimlane/core/resources/app_revision.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import pendulum
+
+from swimlane.core.resources.app import App
+from swimlane.core.resources.base import APIResource
+from swimlane.core.resources.usergroup import UserGroup
+
+
+class AppRevision(APIResource):
```
Comment (nick-allen) [inline swimlane/core/resources/app_revision.py]: Like the approach and expansion here, only concern is that the `AppRevision`/`App` + `RecordRevision`/`Record` are effectively the same things, but have no common inheritance. Seems like quite a bit of copy-pasta in these Revision classes.

Seems like if we really want caching support for these revisions, these classes could just extend the base classes, or better yet find a way to make those base classes respect their revision info if possible.

File: swimlane/core/resources/app_revision.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    @staticmethod
+    def __separator():
+        """
+        Separator for unique ids. Unlikely to be found in application ids. Although technically we do not currently
+        validate app ids in the backend for specific characters so this sequence could be found.
+        """
+        return ' --- '
+
+    @staticmethod
```
Comment (nick-allen) [inline swimlane/core/resources/app_revision.py]: Don't see any reason other than unit tests to have these be static or public

File: swimlane/core/resources/app_revision.py (hunk 1 above)
Comment (jondavidford) [reply swimlane/core/resources/app_revision.py]: Good points! I was thinking about adding a base class for Revisions but it slipped my mind. Do you think adding a revision base class would eliminate the duplication here?

It sounds like you may also be suggesting that Revisions should inherit from App / Record. On the backend models, a Revision contains a App/Record rather than inheriting from it so I think that makes sense here as well. At the moment I don't see how extending App/Record on the Revision classes would simplify anything but I'm open to it.

In terms of making base classes respect their revision info, could you expand on that idea? I'm not quite sure what you mean there.

Comment (nick-allen) [reply swimlane/core/resources/app_revision.py]: Yep, I was suggesting ideally looking for option that keeps a single `App` + `Record` class, and have the revision adapters you added use those instead, modifying those two classes to look at the raw data to see if they are revisions or the current active version. Would have to dig in a bit to see where the shortest option for that is, just seemed like with as much as was copied in, they didn't seem to diverge all that much at first glance.

File: swimlane/core/resources/app_revision.py (hunk 2 above)
Comment (jondavidford) [reply swimlane/core/resources/app_revision.py]: I'm currently using these static methods in the AppRevisionAdapter. Maybe they should be moved there, but then the AppRevision will have to call a method on the Adapter to get its unique id for get_cache_index_keys. Not sure which is cleaner but both seem off to me, can you think of a better way?

Comment (nick-allen) [reply swimlane/core/resources/app_revision.py]: Looking back at the way that caching is looked up, it was definitely written to expect the one arg input, so makes sense how you were trying to patch in the revision number to get it to respect that. 

I'm thinking that'd be the right place to look to fix that in order for the `App`/`AppRevision` classes to be merged and still respect the different version numbers, would think that cache lookup call could be updated to respect combinations of keys, but this is starting to feel overcomplicated.

File: swimlane/core/resources/app_revision.py (hunk 1 above)
Comment (jondavidford) [reply swimlane/core/resources/app_revision.py]: Just committed a base Revision class which should take care of duplication between RecordRevision and AppRevision. What do you see that is duplicated between App and AppRevision, or Record and RecordRevision? For App and AppRevision I see no shared attributes or methods. For Record and RecordRevision, they both have an `app` attribute but one is private. Not seeing how basing AppRevision or RecordRevision on App/Record would simplify things yet?

File: swimlane/core/resources/app_revision.py (hunk 2 above)
Comment (jondavidford) [reply swimlane/core/resources/app_revision.py]: I agree that adjusting the cache implementation would be the best way to clean this up. But might be a bit overkill for now. Maybe best to just get rid of the caching hack here and add it in later if it's needed.

File: swimlane/core/resources/revision_base.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+        self.modified_date = pendulum.parse(self._raw['modifiedDate'])
+        self.revision_number = self._raw['revisionNumber']
+        self.status = self._raw['status']
+
+        # UserGroupSelection, can't set as User without additional lookup
+        self.user = UserGroup(self._swimlane, self._raw['userId'])
+
+        self._raw_version = self._raw['version']
+        self._version = None
```
Comment (jondavidford) [inline swimlane/core/resources/revision_base.py]: The way I am doing this property seems wrong to me. I tried to do `__version` but then could not access from the child classes. Is there a better way to do this?

Comment (cjkoral) [reply swimlane/core/resources/revision_base.py]: here is why you cant access `self.__version` from outside the class https://stackoverflow.com/questions/797771/python-protected-attributes
TLDR; the var is renamed to `_Class__variable` 

thats pythons way of doing private vars, if you need to access outside the class at all the single underscore is your best bet

File: swimlane/core/resources/app_revision.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    Attributes:
+        Attributes:
+        modified_date: The date this app revision was created.
+        revision_number: The revision number of this app revision.
+        status: Indicates whether this revision is the current revision or a historical revision.
+        user: The user that saved this revision of the record.
+        version: The App corresponding to the data contained in this app revision.
+    """
+    def __init__(self, swimlane, raw):
+        super(AppRevision, self).__init__(swimlane, raw)
```
Comment (nick-allen) [inline swimlane/core/resources/app_revision.py]: Redundant `__init__`/`super`, can drop

File: swimlane/core/resources/app_revision.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    @staticmethod
+    def __separator():
+        """
+        Separator for unique ids. Unlikely to be found in application ids. Although technically we do not currently
+        validate app ids in the backend for specific characters so this sequence could be found.
+        """
+        return ' --- '
```
Comment (nick-allen) [inline swimlane/core/resources/app_revision.py]: could just convert to constant attribute at root of class, don't see any reason to have it be a method.

File: swimlane/core/resources/revision_base.py (hunk 3 above)
Comment (nick-allen) [reply swimlane/core/resources/revision_base.py]: https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-a-single-and-a-double-underscore-before-an-object-name/12629901#12629901

Comment (jondavidford) [reply swimlane/core/resources/revision_base.py]: Thanks for the links, I think I understand this better now. It looks to me that since I want to access this in the child classes, single underscore like I have now is the cleanest way to go? Or was there something else you had in mind?
